src/main/kotlin/org/rust/lang/core/types/ty/TyTypeParameter.kt

```diff
@@ -34,7 +36,27 @@ class TyTypeParameter private constructor(
         bounds.flatMap { it.flattenHierarchy }
 
     override fun canUnifyWith(other: Ty, project: Project, mapping: TypeMapping?): Boolean {
```

Comment:
I am not entirely sure, but it may be the case that we want two different `canUnifyWith`: one which is run during type infernece and which can mutate the type mapping, and the other which is run during `impl` search, which does not actually update anything.